Flink: FLIP-27 source split and reader

stevenzwu opened the pull request:
This is the first feature PR for FLIP-27 source split from the uber PR #2105 

Currently, there are at least two open questions to be addressed. Since I will be out for the rest of the week, I like to put this out first.

1) @openinx suggested that we break the `DataIterator` into two levels (combined and file tasks). I have a question that maybe @openinx can confirm [in the comment from the uber PR](https://github.com/apache/iceberg/pull/2105#discussion_r568303231).

2) Reader is currently implemented on top of `FileSourceSplit` and `BulkFormat`. The original reason is that Jingsong mentioned that we may be able to take advantage of the high-performant vectorized readers from Flink. But I am revisiting that decision. It is unlikely Flink's vectorized readers will support deletes. It seems that Iceberg is also adding vectorized readers and I assume Iceberg implementations will support deletes.

@openinx @sundargates @tweise

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/IcebergSourceOptions.java:
```diff
@@ -0,0 +1,33 @@
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.configuration.ConfigOptions;
+
+public interface IcebergSourceOptions {
```
We've introduced a `FlinkTableOptions` , I think it's not friendly to create options classes for source, sink, table etc.  Maybe we could rename the `FlinkTableOptions` to `FlinkConfigOptions`, and put all the options into that class.

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/FlinkBulkFormatAdaptor.java:
```diff
@@ -0,0 +1,222 @@
+
+  @Override
+  public Reader<T> createReader(Configuration config, IcebergSourceSplit split) throws IOException {
+    return new ReaderAdaptor<T>(bulkFormatProvider, config, split, false);
```
Nit:  `new ReaderAdaptor<>(...)` ?

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/FlinkBulkFormatAdaptor.java:
```diff
@@ -0,0 +1,222 @@
+
+  @Override
+  public Reader<T> restoreReader(Configuration config, IcebergSourceSplit split) throws IOException {
+    return new ReaderAdaptor<T>(bulkFormatProvider, config, split, true);
```
ditto

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/FlinkBulkFormatAdaptor.java:
```diff
@@ -0,0 +1,222 @@
+      final CheckpointedPosition position = icebergSplit.checkpointedPosition();
+      if (position != null) {
+        // skip files based on offset in checkpointed position
+        Preconditions.checkArgument(position.getOffset() < icebergSplit.task().files().size(),
```
Nit:  use the method 

```java
checkArgument(boolean expression, @Nullable String errorMessageTemplate, @Nullable Object... errorMessageArgs)
```

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/FlinkBulkFormatAdaptor.java:
```diff
@@ -0,0 +1,222 @@
+        fileOffset++;
+        final FileSourceSplit fileSourceSplit = new FileSourceSplit(
+            "",
+            new Path(URI.create(fileScanTask.file().path().toString())),
```
So the  `FileSourceSplit` will use the flink's fs interface to access the underlying files ?   We iceberg currently has our own `FileIO` interface,  the object storage services are implementing this interface to write/read data to cloud.  If we introduce flink fs here ,  I'm concerning that we have to implement both flink fs interfaces and iceberg FileIO interfaces for making the experimental unified source work.

stevenzwu replied:
We extended from `FileSourceSplit` mainly for the `BulkFormat` batch reader interface so that we can plug in vectorized readers from Flink. I am also debating if this is the right thing to do as mentioned in the original description.

But this is not really relevant to FileIO, which deals with underneath filesystem like S3. Here we are mainly talking about file format reader (like Parquet, Orc).

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/FlinkBulkFormatAdaptor.java:
```diff
@@ -0,0 +1,222 @@
+
+    private final long fileOffset;
+    private final RecordIterator<T> iterator;
+    private final MutableRecordAndPosition mutableRecordAndPosition;
```
Nit:  use `MutableRecordAndPosition<T>` here

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/IcebergSourceReader.java:
```diff
@@ -0,0 +1,74 @@
+import org.apache.iceberg.flink.source.split.MutableIcebergSourceSplit;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+public class IcebergSourceReader<T> extends
```
This class don't have to be introduced in this PR ? I see there's no usage.

stevenzwu replied:
Yes, it is not used in this PR. It is added for the completeness of the `split reader` module

holdenk replied:
Lets drop it for now and do it in a follow up PR, this one is already really really large.

stevenzwu replied:
sounds good. will do

stevenzwu commented:
@openinx I updated the PR on refactoring the DataIterator: using composition instead of inheritance (like the old RowDataIterator). Please take a look when you got a chance.

stevenzwu commented:
@openinx if we are committed to have vectorized reader in `iceberg-flink` that @zhangjun0x01 is working on (PR #2566 ), then I will update this PR and avoid the dep/extension from `FileSourceSplit` and `BulkFormat`. The only reason I did it is to reuse the vectorized readers from Flink.

I am also wondering if the vectorized readers support deletes filtering? I know Flink vectorized readers impl for sure won't support as it is an Iceberg concept.

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/source/CombinedScanTaskIterator.java:
```diff
@@ -0,0 +1,160 @@
+    return position;
+  }
+
+  public static class Position {
```
@openinx following up your comment from the uber PR: https://github.com/apache/iceberg/pull/2105/files#r630834205.

The reason I introduce this mutable `Position` class is to avoid the construction of a <fileOffset, recordOffset> object. It is the current cursor for the iterator. 

Didn't track the recordOffset inside the `FileIteratorReader` for the same reason. Otherwise, `position()` getter will construct a new object each time.

We can't use CheckpointedPosition from Flink for two reasons: (1) it is immutable (2) we want to return the current position (not necessarily CheckpointedPosition).

holdenk commented:
Hey @stevenzwu can you rebase or merge in master branch github is showing a conflicting file.

holdenk commented on flink/src/main/java/org/apache/iceberg/flink/FlinkConfigOptions.java:
```diff
@@ -23,9 +23,9 @@
 import org.apache.flink.configuration.ConfigOptions;
 
-public class FlinkTableOptions {
+public class FlinkConfigOptions {
```
This looks like an existing public API, renaming might have some challenges or at the very least need a comment in the migration guide. Another option would be to extend FlinkConfigOptions into a deprecated FlinkTableOptions so as not to break compatiblity.

stevenzwu replied:
the class is use internally. the config keys are public contract and weren't be changed. Let me move this one out to a separate small PR

stevenzwu replied:
https://github.com/apache/iceberg/pull/2742

holdenk commented on flink/src/main/java/org/apache/iceberg/flink/source/ScanContext.java:
```diff
@@ -34,7 +34,7 @@
  * Context object with optional arguments for a Flink Scan.
  */
-class ScanContext implements Serializable {
+public class ScanContext implements Serializable {
```
Why do we need to make so much more of this ScanContext public?

stevenzwu replied:
because now this class is accessed by classes in sub package (like `reader/RowDataIteratorBulkFormat.java`)

holdenk reviewed (commented):
Thanks for working on this. I don't have enough context around this I'll try and spin up some more context for next week but in the meantime if any of the people with more context (like @JingsongLi or @rdblue ) have some cycles to review that would be rad :)

stevenzwu commented:
@holdenk thx a lot for taking a look. @rdblue @JingsongLi please help take a look and move the new FLIP-27 based Iceberg source move forward if you can. It is part 3 of this uber PR #2105

Right now, there is a pending decision. Currently, this PR is based on the premise of reusing the BulkFormat from Flink for vectorized readers (for Parquet, Orc etc.), originally suggested by @JingsongLi . I am rethinking that choice. It is unlikely Flink (vectorized) readers with support delete filters like Iceberg readers. Maybe iceberg-flink module needs to have its own vectorized readers to support deletes. @zhangjun0x01 already submitted a PR #2566 for Orc. Then this PR needs to be adjusted and break away from flink file source.

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/source/split/IcebergSourceSplit.java:
```diff
@@ -0,0 +1,129 @@
+
+public class IcebergSourceSplit extends FileSourceSplit {
+
+  public enum Status {
```
make sense. will move it out

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/source/split/IcebergSourceSplit.java:
```diff
@@ -0,0 +1,129 @@
+import org.apache.iceberg.relocated.com.google.common.base.Objects;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+
+public class IcebergSourceSplit extends FileSourceSplit {
```
Was hoping to extend from flink-connector-files to leverage the vectorized readers in Flink. I have been revisiting that decision. After discussion with @openinx offline, we agreed that iceberg-flink probably should have its own vectorized readers (mainly for delete filters in the future). flink-connector-files' vectorized readers won't support delete filter, as it is a concept for Iceberg only (not the raw file format like Parquet or Orc). Will update this PR

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/source/split/IcebergSourceSplit.java:
```diff
@@ -0,0 +1,129 @@
+    return checkpointedPosition;
+  }
+
+  public byte[] serializedFormCache() {
```
it is actually used by `IcebergSourceSplit`

tweise replied:
You mean by `IcebergSourceSplitSerializer`. Since that's same package it doesn't need to be public?

stevenzwu commented:
@openinx @sundargates @holdenk @tweise Now this sub PR is ready for review. 

I have changed the code not to extend from FileSourceSplit and BulkFormat, as we are aligned that Iceberg source reader probably can't reuse the vectorized readers from Flink. The main reason is that future Iceberg V2 format supports deletes, which is a concept applicable to Iceberg (not the raw file formats like Orc, Parquet etc.). Hence we can't reuse Flink's vectorized readers with delete filters.

I also moved some unrelated refactoring out of this PR.

sundargates commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/ReaderFactory.java:
```diff
@@ -0,0 +1,102 @@
+   *
+   * @param <T> output record type
+   */
+  interface Reader<T> {
```
Would it make sense to have the reader extend CloseableIterable<RecordsWithSplitIds<T>>? This way the intermediate DS called FileRecords can be completely avoided. A lot of the abstraction makes it slightly complex to read otherwise IMO. 

```
interface Reader<T> extends CloseableIterable<RecordsWithSplitIds<T>> {
}
```

stevenzwu replied:
I would need `FileRecords`, as it is an implementation of the `RecordsWithSplitIds` interface

sundargates commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/ReaderFactory.java:
```diff
@@ -0,0 +1,102 @@
+   *
+   * @param <T> The type of the record.
+   */
+  interface RecordIterator<T> {
```
I'm not sure we need this internal interface given RecordsWithSplitIds is pretty much a superset of this DS.

sundargates commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/RowDataIteratorReaderFactory.java:
```diff
@@ -0,0 +1,138 @@
+import org.apache.iceberg.flink.source.ScanContext;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+
+public class RowDataIteratorReaderFactory implements ReaderFactory<RowData> {
```
Can we make this generic by asking the user to provide the factory for generating T (RowData in this specific case) from a given CombinedScanTask or make that abstract and have a default implementation for RowData that uses the DataIterator?

```
class BatchDataIteratorFactory<T> implements Function<FileScanTaskSplit, CloseableIterable<RecordsWithSplitIds<T>> {
  protected abstract DataIterator<T> getIteratorFor(CombinedScanTask task);
}
```

sundargates commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/ReaderFactory.java:
```diff
@@ -0,0 +1,102 @@
+   * @param split  Iceberg source split
+   * @return a batch reader
+   */
+  Reader<T> create(Configuration config, IcebergSourceSplit split);
```
Why is the configuration passed per split? Wouldn't it be the same for the full table? If so, should it be passed to the constructor as a property of the implementation rather than per split?

stevenzwu replied:
make sense. will change

sundargates commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/ReaderFactory.java:
```diff
@@ -0,0 +1,102 @@
+import org.apache.flink.connector.file.src.util.RecordAndPosition;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+
+public interface ReaderFactory<T> extends Serializable {
```
I think it might be better to avoid this interface and replace it with an existing Java type such as java.util.Function,  as it would lead to less need for understanding new types. If you want to have this abstraction, then you can just define this interface as 
```
public interface ReaderFactory<T> extends Function<IcebergSourceSplit, CloseableIterable<RecordsWithSplitIds<T>> {}
```

stevenzwu commented on build.gradle:
```diff
@@ -140,10 +140,11 @@ subprojects {
       void onOutput(TestDescriptor testDescriptor, TestOutputEvent testOutputEvent) {
         if (lastDescriptor != testDescriptor) {
-          buildLog << "--------\n- Test log for: "<< testDescriptor << "\n--------\n"
+          buildLog << "--------\n- Test log for: " << testDescriptor << "\n--------\n"
```
yeah. annoying auto-formatting from Intellij. will fix it.

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/FlinkConfigOptions.java:
```diff
@@ -40,4 +40,10 @@ private FlinkConfigOptions() {
+  public static final ConfigOption<Integer> SOURCE_READER_FETCH_BATCH_SIZE = ConfigOptions
+      .key("source.iceberg.reader.fetch-batch-size")
+      .intType()
+      .defaultValue(2048)
```
I copied this default value from Flink. I don't actually know what would be the good default value here. We probably need to test it and update accordingly.

```
public class VectorizedColumnBatch implements Serializable {
    /**
     * This number is carefully chosen to minimize overhead and typically allows one
     * VectorizedColumnBatch to fit in cache.
     */
    public static final int DEFAULT_SIZE = 2048;
```

openinx replied:
It's byte size , or row size (I guess it's row-size) ?  From the name,  I did not get the meaning at the first glance. Maybe we will need a more clear name ?

tweise replied:
Or specify the unit in the description.

stevenzwu replied:
updated description with `The target number of records ...`

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java:
```diff
@@ -114,4 +145,43 @@ public void close() throws IOException {
     tasks = null;
   }
+
+  public Position position() {
```
That is a good point. It is also related to your question above. Let me see how to unify them and maybe move away from Flink's CheckpointedPosition

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/FileRecords.java:
```diff
@@ -0,0 +1,92 @@
+public class FileRecords<T> implements RecordsWithSplitIds<RecordAndPosition<T>> {
+
+  @Nullable
+  private final ReaderFactory.RecordIterator recordsForSplit;
```
they should be. will fix

stevenzwu commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/DataIteratorReaderFactory.java:
```diff
@@ -0,0 +1,50 @@
+  private final Configuration config;
+  private final DataIteratorBatcher<T> batcher;
+
+  public DataIteratorReaderFactory(Configuration config, DataIteratorBatcher<T> batcher) {
```
@sundargates I introduced this Batcher interface. RowDataIterator reuses RowData object so that we need to use array pool for the reader. Avro iterator doesn't reuse object. So we don't need to use array pool and clone object. A different Batcher can be plugged in for Avro record

openinx commented on build.gradle:
```diff
@@ -320,6 +321,7 @@ project(':iceberg-flink') {
     compile project(':iceberg-parquet')
     compile project(':iceberg-hive-metastore')
 
+    compileOnly "org.apache.flink:flink-connector-base"
```
The interfaces (such as `SplitReader`, `RecordsWithSplitIds`, `SourceReaderOptions`) from this jar  are stable enough to expose for downstream project like iceberg , ha ?   I rise this Q because I don't see it's marked with a `Public` annotation and seems don't have any compatibility guarantee ?

stevenzwu replied:
`SplitReader`, `RecordsWithSplitIds` are part of the core API that source implementation needs to extend/implement from.

stevenzwu replied:
I did a little research and saw Stephan recently merged the `@PublicEvolve` annotation to all 3 interfaces that you pointed out: https://issues.apache.org/jira/browse/FLINK-22358

tweise replied:
It would be too restrictive to only depend on `@Public` interfaces. FLIP-27 connectors are becoming the default and these interface should mature soon though. Iceberg project will need a story to support multiple Flink versions at some point.

openinx commented on build.gradle:
```diff
@@ -141,6 +141,7 @@ subprojects {
     def buildLog = new File(logFile)
     addTestOutputListener(new TestOutputListener() {
       def lastDescriptor
+
```
Nit:  please don't introduce any unrelated changes for avoiding conflicts when rebase to people's own repo.

stevenzwu replied:
sorry. that was a mistake. will fix

openinx commented on build.gradle:
```diff
@@ -406,6 +409,8 @@ project(':iceberg-flink-runtime') {
       exclude group: 'com.google.code.findbugs', module: 'jsr305'
     }
+
+    compile "org.apache.flink:flink-connector-base"
```
This jar was not included into the flink-dist binary package, so we have to include it into iceberg-flink-runtime jar explicitly ?

stevenzwu replied:
that is correct

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/FlinkSplitGenerator.java:
```diff
@@ -27,23 +27,39 @@
+    }
+  }
+
+  public static List<IcebergSourceSplit> planIcebergSourceSplits(
```
Should we add a javadoc to indicate why do we need to add an extra `planIcebergSourceSplits` (compared to `createInputSplits`) ?  I think it's used for implementing the flip27's SourceSplit, right ?

openinx replied:
It's good to align with the `createInputSplits` by naming this as `createIcebergSourceSplits`.

stevenzwu replied:
actually, I think we should rename `createInputSplits` to `planFlinkInputSplit`. We are not creating splits out of nowhere. Both are just discover/plan splits from table and call the same `planTasks`. I can add some javadoc on the new `planIcebergSourceSplits` method.

Since `createInputSplits` is non-public, we should be safe to rename.

tweise commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/DataIteratorBatcher.java:
```diff
@@ -0,0 +1,31 @@
+import org.apache.iceberg.flink.source.DataIterator;
+import org.apache.iceberg.io.CloseableIterator;
+
+@FunctionalInterface
```
javadoc?

stevenzwu replied:
will add

tweise commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/DataIteratorReaderFunction.java:
```diff
@@ -0,0 +1,46 @@
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+import org.apache.iceberg.io.CloseableIterator;
+
+public abstract class DataIteratorReaderFunction<T> implements ReaderFunction<T> {
```
internal or javadoc?

stevenzwu replied:
will add Javadoc. In this PR, there is only one implementation of `RowDataReaderFunction`. but we can extend from it with `AvroGenericRecordReaderFunction` that directly reads Parquet files into Avro GenericRecord. That is why this is left as `public`

tweise commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/FileRecords.java:
```diff
@@ -0,0 +1,103 @@
+ * A batch of recrods for one split
+ */
+@Internal
+public class FileRecords<T> implements RecordsWithSplitIds<RecordAndPosition<T>> {
```
Found the class name not intuitive. From usage, this appears to be a "fetch result"?

stevenzwu replied:
I am going to rename it to `SplitRecords`. Javadoc explains the purpose of this class. yes, it is for "fetch result"

tweise commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/IcebergSourceSplitReader.java:
```diff
@@ -0,0 +1,118 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class IcebergSourceSplitReader<T> implements SplitReader<RecordAndPosition<T>, IcebergSourceSplit> {
```
Did you consider the need to subclass the reader for customization? Maybe it should be protected?

stevenzwu replied:
right now, I don't anticipate any need of extending from this class

tweise commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/RecyclableArrayIterator.java:
```diff
@@ -0,0 +1,84 @@
+  }
+
+  /**
+   * Each record's {@link RecordAndPosition} will have the same fileOffset (for {@link RecordAndPosition#getOffset()}.
```
should this move to class level doc?

stevenzwu replied:
will move

tweise commented on flink/src/main/java/org/apache/iceberg/flink/source/split/IcebergSourceSplitSerializer.java:
```diff
@@ -0,0 +1,74 @@
+
+/**
+ * TODO: use Java serialization for now.
+ * will switch to more stable serializer from issue-1698.
```
Please add full link. With new version such subsequent change will be backward compatible?

stevenzwu replied:
will add link.

Regarding the future change with stable serialize, it will be backward compatible as we can bump up the serializer version to v2 and can continue to deserialize state written by v1

```
  @Override
  public IcebergSourceSplit deserialize(int version, byte[] serialized) throws IOException {
    switch (version) {
      case 1:
        return deserializeV1(serialized);
      default:
        throw new IOException("Unknown version: " + version);
    }
  }
```

rdblue commented:
@stevenzwu, I'll try to review this in the next week. Thank you!

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/source/FlinkSplitPlanner.java:
```diff
@@ -22,33 +22,58 @@
-    FlinkInputSplit[] splits = new FlinkInputSplit[tasks.size()];
-    for (int i = 0; i < tasks.size(); i++) {
-      splits[i] = new FlinkInputSplit(i, tasks.get(i));
+  static FlinkInputSplit[] planInputSplits(Table table, ScanContext context) {
```
Why change the name of this method?

stevenzwu replied:
`create/generate` implies creating sth new. This is actually plan/discover splits from table. Hence changed the method name. I actually also renamed the class name from `FlinkSplitGenerator` to `FlinkSplitPlanner`. This is an internal class. So it shouldn't break user code.

stevenzwu replied:
`create/generate` implies creating sth new. This is actually plan/discover splits from table. Hence changed the method name. I actually also renamed the class name from `FlinkSplitGenerator` to `FlinkSplitPlanner`. This is an internal class. So it shouldn't break user code.

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java:
```diff
@@ -41,16 +42,45 @@
+    Preconditions.checkArgument(startingPosition.fileOffset() < combinedTask.files().size(),
+        "Checkpointed file offset is %d, while CombinedScanTask has %d files",
+        startingPosition.fileOffset(), combinedTask.files().size());
+    for (long i = 0L; i < startingPosition.fileOffset(); ++i) {
```
Is `fileOffset()` a `long`? That seems odd to me. When would you need to address more than 2 billion files in a single combined scan task?

stevenzwu replied:
integer would certainly be sufficient. I was using `long` to match the type in `RecordAndPosition` from flink-connector-files module. I have already tried to move away from depending on code from flink-connector-files. This looks like a miss due to the flink-table-planner-blink_2.12 pulling in flink-connector-files transitively.

The `long offset` from Flink's `RecordAndPosition` actually meant byte offset within a file. I will define our own `RecordAndPosition` and change `fileOffset` to `int` type.

stevenzwu replied:
same as another comment. will update

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java:
```diff
@@ -41,16 +42,45 @@
     this.inputFilesDecryptor = new InputFilesDecryptor(task, io, encryption);
+    this.combinedTask = task;
+
     this.tasks = task.files().iterator();
```
I'd probably rename this to `fileTasks` since there is now a `combinedTask`.

stevenzwu replied:
agree. will rename it to `fileTasksIterator`

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java:
```diff
@@ -41,16 +42,45 @@
+            startingPosition.recordOffset());
+      }
+    }
+    this.position.update(startingPosition.fileOffset(), startingPosition.recordOffset());
```
Can `position` be final since this is using `update`?

stevenzwu replied:
yes, will change `Position` to final

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/source/RowDataFileScanTaskReader.java:
```diff
@@ -48,7 +48,6 @@
 
 @Internal
 public class RowDataFileScanTaskReader implements FileScanTaskReader<RowData> {
-
```
Looks like this file doesn't need to change. Can you remove this?

stevenzwu replied:
will revert

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/source/ScanContext.java:
```diff
@@ -68,6 +68,9 @@
       ConfigOptions.key("monitor-interval").durationType().defaultValue(Duration.ofSeconds(10));
 
+  private static final ConfigOption<Boolean> INCLUDE_COLUMN_STATS =
+      ConfigOptions.key("include-column-stats").booleanType().defaultValue(false);
```
How is this addition related to FLIP-27?

stevenzwu replied:
this is needed for event time aligned assigner / rough ordering where we use the min-max stats from timestamp column to order the splits and assignment. it is not directly related to the reader part. It is just part of the sub PR as I copied the classes from the the uber PR #2105 for this sub PR.

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/source/StreamingMonitorFunction.java:
```diff
@@ -140,7 +140,7 @@ private void monitorAndForwardSplits() {
       }
 
-      FlinkInputSplit[] splits = FlinkSplitGenerator.createInputSplits(table, newScanContext);
+      FlinkInputSplit[] splits = FlinkSplitPlanner.planInputSplits(table, newScanContext);
```
Does this keep the monitor function so that it can maintain the old API? Do we need to maintain the old API?

stevenzwu replied:
FlinkSplitGenerator/FlinkSplitPlanner is an internal class. it doesn't affect the public API for user code

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/ArrayPoolDataIteratorBatcher.java:
```diff
@@ -0,0 +1,116 @@
+import org.apache.iceberg.flink.source.Position;
+import org.apache.iceberg.io.CloseableIterator;
+
+class ArrayPoolDataIteratorBatcher<T> implements DataIteratorBatcher<T> {
```
There isn't much context for me to go on here. Should there be Javadoc to explain what's going on?

stevenzwu replied:
will add Javadoc to explain

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/DataIteratorBatcher.java:
```diff
@@ -0,0 +1,36 @@
+ */
+@FunctionalInterface
+public interface DataIteratorBatcher<T> extends Serializable {
+  CloseableIterator<RecordsWithSplitIds<RecordAndPosition<T>>> apply(String splitId, DataIterator<T> inputIterator);
```
Why name this `apply`? Is there a more specific verb you could use here?

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/ArrayPoolDataIteratorBatcher.java:
```diff
@@ -0,0 +1,116 @@
+      int num = 0;
+      while (inputIterator.hasNext() && num < batchSize) {
+        T nextRecord = inputIterator.next();
+        recordFactory.clone(nextRecord, batch[num]);
```
What is this doing? Is `clone` expensive?

stevenzwu replied:
you are exactly right. will add code comment

rdblue commented on flink/src/main/java/org/apache/iceberg/flink/source/reader/ArrayPoolDataIteratorBatcher.java:
```diff
@@ -0,0 +1,116 @@
+          break;
+        }
+      }
+      if (num == 0) {
```
Could you add newlines between control flow blocks?

stevenzwu replied:
will do

stevenzwu commented on flink-runtime/build.gradle:
```diff
@@ -42,6 +42,8 @@ project(':iceberg-flink-runtime') {
       exclude group: 'com.google.code.findbugs', module: 'jsr305'
     }
+
+    implementation "org.apache.flink:flink-connector-base"
```
added comment

openinx replied:
@stevenzwu  After we got this PR merged: https://github.com/apache/iceberg/pull/3364,  we don't need to introduce a common `iceberg-flink-runtime` for all different flink versions, instead we have a different `iceberg-flink:iceberg-flink-<MAJOR.MINO>-runtime` module for different `<MAJRO.MINOR>` flink releases so that we could build the features on top of the latest flink API. 

You may want to add the transitive dependency `org.apache.flink:flink-connector-base` in[ this line for flink 1.12](https://github.com/apache/iceberg/blob/master/flink/v1.12/build.gradle#L154) , and[ this line for flink 1.13.](https://github.com/apache/iceberg/blob/master/flink/v1.13/build.gradle#L154)

openinx commented on flink/build.gradle:
```diff
@@ -28,6 +28,7 @@ project(':iceberg-flink') {
     implementation project(':iceberg-parquet')
     implementation project(':iceberg-hive-metastore')
 
+    compileOnly "org.apache.flink:flink-connector-base"
```
It's strange the the build for flink 1.12 & flink 1.13 has been passed, because I don't see the same dependency are added to flink 1.12 build.gradle and 1.13 build.gradle.  Maybe I need to check the 1.12's build.gradle again.

stevenzwu replied:
@openinx Maybe follow-up on the other [comment discussion](https://github.com/apache/iceberg/pull/3354/files#r740053509) here.

With the SplitEnumerator API change, looks like I need to put FLIP-27 source in the `v1.13` folder. What should we do with future versions (like 1.14)? do we copy the FLIP-27 source code `v1.13` to `v1.14` folder?

rdblue replied:
@openinx, the tests run against the iceberg-flink module. They aren't present in the 1.12 or 1.13 modules. If you want them to be run for those modules, you'd need to add the source folder like you do for `src/main/java`. If you choose to do that, let's also remove CI for the common module since we don't need to run the tests outside of 1.12 and 1.13 if they are run in those modules.

rdblue replied:
@stevenzwu, I think that copying the parts that change is reasonable. And once we remove support for 1.12, you can move the files back into the common module.

stevenzwu replied:
yeah. that is my plan too. Once 1.12 support is removed, we should be able to move files back to the common module. We just need to be diligent with these efforts.

openinx commented on flink/src/main/java/org/apache/iceberg/flink/FlinkConfigOptions.java:
```diff
@@ -40,4 +40,10 @@ private FlinkConfigOptions() {
           .withDescription("Sets max infer parallelism for source operator.");
+
+  public static final ConfigOption<Integer> SOURCE_READER_FETCH_RECORD_BATCH_SIZE = ConfigOptions
+      .key("source.iceberg.reader.fetch-record-batch-size")
```
Yes,  this is unrelated to the table/SQL execution. Both datastream job and table SQL job are using the same configuration keys.  So I'm okay to keep the current name. 

(In fact, if we don't consider the flink's configuration name,  I'd prefer to name it `iceberg.source.reader.fetch-record-batch-size`. But iceberg is also a kind of flink connector, all the other flip-27 source connector has named it as `source.<connector>.xxx`,  so I think we can follow the naming).

openinx commented on flink/src/main/java/org/apache/iceberg/flink/source/DataIterator.java:
```diff
@@ -41,18 +42,47 @@
+    this.combinedTask = task;
+    // fileOffset starts at -1 because we started
+    // from an empty iterator that is not from the split files.
+    this.position = new Position(-1, 0L);
```
The general `DataIterator` don't use the position or `seek` method to skip tasks or records.  Putting all the flip-27 related logics in the flink common read path does not make sense to me, because every times when I read this class, I need to see which part is related to flip-27, which is the unrelated part. 

I will suggest to introduce a separate SeekableDataIterator to isolate the two code path, I made a simple commit for this: https://github.com/openinx/incubator-iceberg/commit/b08dde86aae0c718d9d72acb347dffb3a836b336, you may want to take a look.

stevenzwu replied:
I wouldn't say that `seek` capability is FLIP-27 specific. If we think `DataIterator` as reading a list of files/splits from `CombinedScanTask`, it is like a file API where `seek` is pretty common. It is needed to achieve exactly-once semantics on source reading. 

Thanks a lot for the `SeekableDataIterator`. I feel that leaving these two empty abstract methods in the base `DataIterator` is a little weird
```
protected void advanceRecord()
protected void advanceTask()
```

Overall, I still think adding `seek` capability to `DataIterator` is natural (for file-like read APIs)

stevenzwu commented:
closing this PR for now. Will further break it down to smaller PR as suggested by @openinx
